README.md
![flutter-native-scaffold-intro](https://user-images.githubusercontent.com/102783954/235310805-89819d0c-617f-41f8-8795-681cd98faa41.png)

## How to get started❓ 🤔

You can simply add the dependency from **cli** like:
`flutter pub add native_scaffold`

**Or** you can add it in your **pubspec.yaml** like:

```yaml
dependencies:
  native_scaffold:
```

### What is different?

| Title                                                        | Scaffold | CustomScaffold |
| ------------------------------------------------------------ | -------- | -------------- |
| Gives you basic AppBar and Scaffold functionalities          | ✅       | ✅             |
| Easy to use and setup                                        | 👎🏻       | ✅             |
| Works differently for a better experiance in Android and iOS | 👎🏻       | ✅             |
| Easy to add sliver scroll animation                          | 👎🏻       | ✅             |
| Allows you to add a bottom floating bar                      | 👎🏻       | ✅             |
| Can easily use `onRefresh` function                          | 👎🏻       | ✅             |

## How to use❓ 🤔

Using it is as **easy as** create a **Container**, Don't beleave? Let me show you:

```dart
NativeScaffold(
	title: "This is easy",
	child: ...,
);
```

## Additional Params (Optional)

- `title`: The title for the app bar or navigation bar.

- `action`: An action widget displayed at the end of the app bar or navigation bar.

- `leading`: A leading widget displayed at the start of the app bar or navigation bar.

- `child`: The main content of the scaffold.

- `showLoargeTitleInIOS`: Whether to show a large title on iOS. This shall only be used on iOS.

- `sliverAppBarPinned`: Whether to pin the SliverAppBar in place when scrolling. This shall only be used on Android.

- `onRefresh`: A function to be called when the refresh gesture is triggered.

- `sliverAppBarWidget`: A custom SliverAppBar widget. This shall only be used on Android.

- `appBarCenterWidget`: A widget to be displayed in the center of the app bar.

- `sliverAppBarWidgetHeight`: The height of the custom SliverAppBar widget.

- `bottomFloatingBar`: A widget to be used as the bottom navigation bar.

- `backgroundColor`: The background color of the scaffold.

- `appBarEnabled`: Whether the app bar should be enabled.

**This is what Zenzzen Team made for you specially 😊, Now it's your responsibility to make ground-breaking app that rocks 🤘🏻**

# Some Apps Made Using This 😎

![other-apps-using-custom-scaffold](https://user-images.githubusercontent.com/102783954/235311050-3ee430e3-88dd-4b0d-900a-cb97f37692f4.png)


CHANGELOG.md
# 1.0.0

The boom begans

# 1.0.1

Added Example

# 1.0.2

Bug Fix


lib/native_scaffold.dart
import 'dart:io';

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

/// A native scaffold widget that adapts to different platforms (iOS and Android).
class NativeScaffold extends StatelessWidget {
  /// The title for the app bar or navigation bar.
  final String title;

  /// An action widget displayed at the end of the app bar or navigation bar.
  final Widget? action;

  /// A leading widget displayed at the start of the app bar or navigation bar.
  final Widget? leading;

  /// The main content of the scaffold.
  final Widget child;

  /// Whether to show a large title on iOS. This shall only be used on iOS.
  final bool showLoargeTitleInIOS;

  /// Whether to pin the SliverAppBar in place when scrolling. This shall only be used on Android.
  final bool sliverAppBarPinned;

  /// A function to be called when the refresh gesture is triggered.
  final Function? onRefresh;

  /// A custom SliverAppBar widget. This shall only be used on Android.
  final Widget? sliverAppBarWidget;

  /// A widget to be displayed in the center of the app bar.
  final Widget? appBarCenterWidget;

  /// The height of the custom SliverAppBar widget.
  final double? sliverAppBarWidgetHeight;

  /// A widget to be used as the bottom navigation bar.
  final Widget? bottomFloatingBar;

  /// The background color of the scaffold.
  final Color? backgroundColor;

  /// Whether the app bar should be enabled.
  final bool? appBarEnabled;

  /// Creates a native scaffold widget.
  const NativeScaffold({
    Key? key,
    required this.title,
    this.action,
    this.leading,
    required this.child,
    this.showLoargeTitleInIOS = false,
    this.sliverAppBarPinned = true,
    this.onRefresh,
    this.sliverAppBarWidget,
    this.sliverAppBarWidgetHeight = 160.0,
    this.bottomFloatingBar,
    this.backgroundColor,
    this.appBarCenterWidget,
    this.appBarEnabled = true,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Platform.isIOS
        ? CupertinoScaffold(
            backgroundColor: backgroundColor,
            showLoargeTitleInIOS: showLoargeTitleInIOS,
            appBarEnabled: appBarEnabled,
            appBarCenterWidget: appBarCenterWidget,
            title: title,
            action: action,
            leading: leading,
            onRefresh: onRefresh,
            sliverAppBarWidget: sliverAppBarWidget,
            bottomFloatingBar: bottomFloatingBar,
            child: child,
          )
        : AndroidScaffold(
            backgroundColor: backgroundColor,
            sliverAppBarWidget: sliverAppBarWidget,
            appBarEnabled: appBarEnabled,
            appBarCenterWidget: appBarCenterWidget,
            title: title,
            action: action,
            leading: leading,
            onRefresh: onRefresh,
            sliverAppBarPinned: sliverAppBarPinned,
            sliverAppBarWidgetHeight: sliverAppBarWidgetHeight,
            bottomFloatingBar: bottomFloatingBar,
            child: child,
          );
  }
}

/// A custom Cupertino scaffold widget that adapts to different platforms (iOS).
class CupertinoScaffold extends StatelessWidget {
  /// Creates a Cupertino scaffold widget.
  const CupertinoScaffold({
    Key? key,
    required this.backgroundColor,
    required this.showLoargeTitleInIOS,
    required this.appBarEnabled,
    required this.appBarCenterWidget,
    required this.title,
    required this.action,
    required this.leading,
    required this.onRefresh,
    required this.sliverAppBarWidget,
    required this.child,
    required this.bottomFloatingBar,
  }) : super(key: key);

  /// The background color of the scaffold.
  final Color? backgroundColor;

  /// Whether to show a large title on iOS.
  final bool showLoargeTitleInIOS;

  /// Whether the app bar should be enabled.
  final bool? appBarEnabled;

  /// A widget to be displayed in the center of the app bar.
  final Widget? appBarCenterWidget;

  /// The title for the app bar or navigation bar.
  final String title;

  /// An action widget displayed at the end of the app bar or navigation bar.
  final Widget? action;

  /// A leading widget displayed at the start of the app bar or navigation bar.
  final Widget? leading;

  /// A function to be called when the refresh gesture is triggered.
  final Function? onRefresh;

  /// A custom SliverAppBar widget.
  final Widget? sliverAppBarWidget;

  /// The main content of the scaffold.
  final Widget child;

  /// A widget to be used as the bottom navigation bar.
  final Widget? bottomFloatingBar;

  @override
  Widget build(BuildContext context) {
    return CupertinoPageScaffold(
      backgroundColor: backgroundColor,
      navigationBar: showLoargeTitleInIOS || !appBarEnabled!
          ? null
          : CupertinoNavigationBar(
              border: Border.all(color: Colors.transparent),
              middle: appBarCenterWidget ??
                  Text(
                    title,
                  ),
              trailing: action,
              leading: leading,
            ),
      child: showLoargeTitleInIOS
          ? Stack(
              children: [
                CupertinoScrollbar(
                  child: CustomScrollView(
                    slivers: [
                      !appBarEnabled!
                          ? Container()
                          : CupertinoSliverNavigationBar(
                              largeTitle: Text(title),
                              trailing: action,
                              leading: leading,
                            ),
                      if (onRefresh != null)
                        CupertinoSliverRefreshControl(
                          onRefresh: onRefresh as Future<void> Function()?,
                        ),
                      SliverList(
                        delegate: SliverChildListDelegate(
                          [
                            if (sliverAppBarWidget != null) sliverAppBarWidget!,
                            child,
                          ],
                        ),
                      )
                    ],
                  ),
                ),
                if (bottomFloatingBar != null) bottomFloatingBar!,
              ],
            )
          : SafeArea(
              bottom: false,
              child: Stack(
                children: [
                  CupertinoScrollbar(
                    child: CustomScrollView(
                      slivers: [
                        if (onRefresh != null)
                          CupertinoSliverRefreshControl(
                            onRefresh: onRefresh as Future<void> Function()?,
                          ),
                        SliverFillRemaining(
                          child: CupertinoScrollbar(
                            child: SingleChildScrollView(
                              child: Column(
                                crossAxisAlignment: CrossAxisAlignment.start,
                                children: [
                                  if (sliverAppBarWidget != null)
                                    sliverAppBarWidget!,
                                  child,
                                ],
                              ),
                            ),
                          ),
                        )
                      ],
                    ),
                  ),
                  if (bottomFloatingBar != null) bottomFloatingBar!,
                ],
              ),
            ),
    );
  }
}

/// A custom Android scaffold widget that adapts to different platforms (Android).
class AndroidScaffold extends StatelessWidget {
  /// Creates an Android scaffold widget.
  const AndroidScaffold({
    Key? key,
    required this.backgroundColor,
    required this.sliverAppBarWidget,
    required this.appBarEnabled,
    required this.appBarCenterWidget,
    required this.title,
    required this.action,
    required this.leading,
    required this.onRefresh,
    required this.sliverAppBarPinned,
    required this.sliverAppBarWidgetHeight,
    required this.child,
    required this.bottomFloatingBar,
  }) : super(key: key);

  /// The background color of the scaffold.
  final Color? backgroundColor;

  /// A custom SliverAppBar widget.
  final Widget? sliverAppBarWidget;

  /// Whether the app bar should be enabled.
  final bool? appBarEnabled;

  /// A widget to be displayed in the center of the app bar.
  final Widget? appBarCenterWidget;

  /// The title for the app bar.
  final String title;

  /// An action widget displayed at the end of the app bar.
  final Widget? action;

  /// A leading widget displayed at the start of the app bar.
  final Widget? leading;

  /// A function to be called when the refresh gesture is triggered.
  final Function? onRefresh;

  /// Whether to pin the SliverAppBar in place when scrolling.
  final bool sliverAppBarPinned;

  /// The height of the custom SliverAppBar widget.
  final double? sliverAppBarWidgetHeight;

  /// The main content of the scaffold.
  final Widget child;

  /// A widget to be used as the bottom navigation bar.
  final Widget? bottomFloatingBar;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: backgroundColor ?? Colors.transparent,
      appBar: sliverAppBarWidget != null || !appBarEnabled!
          ? null
          : AppBar(
              title: appBarCenterWidget ??
                  Text(
                    title,
                  ),
              centerTitle: appBarCenterWidget == null ? true : false,
              titleSpacing: appBarCenterWidget == null ? null : 0.0,
              actions: [
                action ?? const SizedBox.shrink(),
              ],
              leading: leading,
            ),
      body: Stack(
        children: [
          CustomScrollView(
            slivers: [
              if (onRefresh != null)
                SliverToBoxAdapter(
                  child: RefreshIndicator(
                    onRefresh: onRefresh as Future<void> Function(),
                    child: const SizedBox.shrink(),
                  ),
                ),
              if (sliverAppBarWidget != null && appBarEnabled!)
                SliverAppBar(
                  pinned: sliverAppBarPinned,
                  snap: false,
                  floating: false,
                  expandedHeight: sliverAppBarWidgetHeight,
                  flexibleSpace: FlexibleSpaceBar(
                    background: sliverAppBarWidget,
                  ),
                  leading: leading,
                  actions: [
                    action ?? const SizedBox.shrink(),
                  ],
                  title: Text(
                    title,
                  ),
                ),
              SliverList(
                delegate: SliverChildBuilderDelegate(
                  (context, index) => child,
                  childCount: 1,
                ),
              ),
            ],
          ),
          if (bottomFloatingBar != null) bottomFloatingBar!,
        ],
      ),
    );
  }
}


example/linux/my_application.cc
#include "my_application.h"

#include <flutter_linux/flutter_linux.h>
#ifdef GDK_WINDOWING_X11
#include <gdk/gdkx.h>
#endif

#include "flutter/generated_plugin_registrant.h"

struct _MyApplication {
  GtkApplication parent_instance;
  char** dart_entrypoint_arguments;
};

G_DEFINE_TYPE(MyApplication, my_application, GTK_TYPE_APPLICATION)

// Implements GApplication::activate.
static void my_application_activate(GApplication* application) {
  MyApplication* self = MY_APPLICATION(application);
  GtkWindow* window =
      GTK_WINDOW(gtk_application_window_new(GTK_APPLICATION(application)));

  // Use a header bar when running in GNOME as this is the common style used
  // by applications and is the setup most users will be using (e.g. Ubuntu
  // desktop).
  // If running on X and not using GNOME then just use a traditional title bar
  // in case the window manager does more exotic layout, e.g. tiling.
  // If running on Wayland assume the header bar will work (may need changing
  // if future cases occur).
  gboolean use_header_bar = TRUE;
#ifdef GDK_WINDOWING_X11
  GdkScreen* screen = gtk_window_get_screen(window);
  if (GDK_IS_X11_SCREEN(screen)) {
    const gchar* wm_name = gdk_x11_screen_get_window_manager_name(screen);
    if (g_strcmp0(wm_name, "GNOME Shell") != 0) {
      use_header_bar = FALSE;
    }
  }
#endif
  if (use_header_bar) {
    GtkHeaderBar* header_bar = GTK_HEADER_BAR(gtk_header_bar_new());
    gtk_widget_show(GTK_WIDGET(header_bar));
    gtk_header_bar_set_title(header_bar, "native_scaffold_example");
    gtk_header_bar_set_show_close_button(header_bar, TRUE);
    gtk_window_set_titlebar(window, GTK_WIDGET(header_bar));
  } else {
    gtk_window_set_title(window, "native_scaffold_example");
  }

  gtk_window_set_default_size(window, 1280, 720);
  gtk_widget_show(GTK_WIDGET(window));

  g_autoptr(FlDartProject) project = fl_dart_project_new();
  fl_dart_project_set_dart_entrypoint_arguments(project, self->dart_entrypoint_arguments);

  FlView* view = fl_view_new(project);
  gtk_widget_show(GTK_WIDGET(view));
  gtk_container_add(GTK_CONTAINER(window), GTK_WIDGET(view));

  fl_register_plugins(FL_PLUGIN_REGISTRY(view));

  gtk_widget_grab_focus(GTK_WIDGET(view));
}

// Implements GApplication::local_command_line.
static gboolean my_application_local_command_line(GApplication* application, gchar*** arguments, int* exit_status) {
  MyApplication* self = MY_APPLICATION(application);
  // Strip out the first argument as it is the binary name.
  self->dart_entrypoint_arguments = g_strdupv(*arguments + 1);

  g_autoptr(GError) error = nullptr;
  if (!g_application_register(application, nullptr, &error)) {
     g_warning("Failed to register: %s", error->message);
     *exit_status = 1;
     return TRUE;
  }

  g_application_activate(application);
  *exit_status = 0;

  return TRUE;
}

// Implements GObject::dispose.
static void my_application_dispose(GObject* object) {
  MyApplication* self = MY_APPLICATION(object);
  g_clear_pointer(&self->dart_entrypoint_arguments, g_strfreev);
  G_OBJECT_CLASS(my_application_parent_class)->dispose(object);
}

static void my_application_class_init(MyApplicationClass* klass) {
  G_APPLICATION_CLASS(klass)->activate = my_application_activate;
  G_APPLICATION_CLASS(klass)->local_command_line = my_application_local_command_line;
  G_OBJECT_CLASS(klass)->dispose = my_application_dispose;
}

static void my_application_init(MyApplication* self) {}

MyApplication* my_application_new() {
  return MY_APPLICATION(g_object_new(my_application_get_type(),
                                     "application-id", APPLICATION_ID,
                                     "flags", G_APPLICATION_NON_UNIQUE,
                                     nullptr));
}


example/windows/runner/main.cpp
#include <flutter/dart_project.h>
#include <flutter/flutter_view_controller.h>
#include <windows.h>

#include "flutter_window.h"
#include "utils.h"

int APIENTRY wWinMain(_In_ HINSTANCE instance, _In_opt_ HINSTANCE prev,
                      _In_ wchar_t *command_line, _In_ int show_command) {
  // Attach to console when present (e.g., 'flutter run') or create a
  // new console when running with a debugger.
  if (!::AttachConsole(ATTACH_PARENT_PROCESS) && ::IsDebuggerPresent()) {
    CreateAndAttachConsole();
  }

  // Initialize COM, so that it is available for use in the library and/or
  // plugins.
  ::CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED);

  flutter::DartProject project(L"data");

  std::vector<std::string> command_line_arguments =
      GetCommandLineArguments();

  project.set_dart_entrypoint_arguments(std::move(command_line_arguments));

  FlutterWindow window(project);
  Win32Window::Point origin(10, 10);
  Win32Window::Size size(1280, 720);
  if (!window.Create(L"native_scaffold_example", origin, size)) {
    return EXIT_FAILURE;
  }
  window.SetQuitOnClose(true);

  ::MSG msg;
  while (::GetMessage(&msg, nullptr, 0, 0)) {
    ::TranslateMessage(&msg);
    ::DispatchMessage(&msg);
  }

  ::CoUninitialize();
  return EXIT_SUCCESS;
}


example/lib/main.dart
// ignore_for_file: prefer_const_literals_to_create_immutables, prefer_const_constructors

import 'package:flutter/material.dart';
import 'package:native_scaffold/native_scaffold.dart';

void main(List<String> args) {
  runApp(MainApp());
}

class MainApp extends StatelessWidget {
  const MainApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        useMaterial3: true,
      ),
      home: NativeScaffold(
        title: "Title",
        onRefresh: () async {
          await Future.delayed(Duration(seconds: 2));
        },
        action: IconButton(
          icon: Icon(Icons.add),
          onPressed: () {},
        ),
        appBarEnabled: true,
        backgroundColor: Colors.white,
        bottomFloatingBar: SizedBox(
          height: double.infinity,
          child: Stack(
            alignment: Alignment.bottomCenter,
            children: [
              Container(
                height: 50,
                width: double.infinity,
                color: Colors.red,
              ),
            ],
          ),
        ),
        sliverAppBarWidget: Image.network(
          "https://user-images.githubusercontent.com/102783954/235310805-89819d0c-617f-41f8-8795-681cd98faa41.png",
          fit: BoxFit.cover,
        ),
        sliverAppBarPinned: true,
        sliverAppBarWidgetHeight: 300,
        showLoargeTitleInIOS: true,
        leading: IconButton(
          icon: Icon(Icons.menu),
          onPressed: () {},
        ),
        appBarCenterWidget: Text("Center"),
        child: Column(
          children: [
            SizedBox(
              height: 2000,
            ),
          ],
        ),
      ),
    );
  }
}
